### ISSUE TITLE
Crash on arm with "A fatal error has been detected by the Java Runtime Environment: SIGILL (0x4)"

### ISSUE THREAD

**Issue by sdelafond** [xerial/sqlite-jdbc#186]
This happens with 3.15.1, running the basic Sample class, on an Asus
ac88u running a Debian armel userland: although this is an armv7
processor, it lacks flags required to run a Debian armhf userland.

In this context, the proper
./org/sqlite/native/Linux/arm/libsqlitejdbc.so is being loaded, and ldd
seems happy with it:

# ldd ./org/sqlite/native/Linux/arm/libsqlitejdbc.so
        linux-vdso.so.1 (0x7efbe000)
        libc.so.6 => /lib/arm-linux-gnueabi/libc.so.6 (0x76cdb000)
        /lib/ld-linux.so.3 (0x54ad6000)

However, I still get this crash:

Stack: [0x76616000,0x76666000],  sp=0x76664614,  free space=313k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [sqlite-3.15.1-12a64d26-7d96-4bb4-a605-483e3c1ce73e-libsqlitejdbc.so+0x3c3fc]

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.sqlite.core.NativeDB._open_utf8([BI)V+0
j  org.sqlite.core.NativeDB._open(Ljava/lang/String;I)V+6
j  org.sqlite.core.DB.open(Lorg/sqlite/SQLiteConnection;Ljava/lang/String;I)V+8
j  org.sqlite.core.CoreConnection.open(II)V+408
j  org.sqlite.core.CoreConnection.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/util/Properties;)V+138
j  org.sqlite.jdbc3.JDBC3Connection.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/util/Properties;)V+4
j  org.sqlite.jdbc4.JDBC4Connection.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/util/Properties;)V+4
j  org.sqlite.SQLiteConnection.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/util/Properties;)V+4
j  org.sqlite.JDBC.createConnection(Ljava/lang/String;Ljava/util/Properties;)Ljava/sql/Connection;+24
j  org.sqlite.JDBC.connect(Ljava/lang/String;Ljava/util/Properties;)Ljava/sql/Connection;+2
j  java.sql.DriverManager.getConnection(Ljava/lang/String;Ljava/util/Properties;Ljava/lang/Class;)Ljava/sql/Connection;+171
j  java.sql.DriverManager.getConnection(Ljava/lang/String;)Ljava/sql/Connection;+13
j  JDBC_test.main([Ljava/lang/String;)V+30
v  ~StubRoutines::call_stub

Here is a complete description of the system:

os.name = Linux
os.arch = arm

# java -version
openjdk version "1.8.0_111"
OpenJDK Runtime Environment (build 1.8.0_111-8u111-b14-2~bpo8+1-b14)
OpenJDK Zero VM (build 25.111-b14, interpreted mode)

# uname -a
Linux test.example.com 4.4.34 #3 SMP Fri Nov 25 15:52:23 CET 2016 armv7l GNU/Linux

# cat /proc/cpuinfo
model name      : ARMv7 Processor rev 0 (v7l)
processor       : 0
BogoMIPS        : 2786.91
Features        : half fastmult edsp tls 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x3
CPU part        : 0xc09
CPU revision    : 0

model name      : ARMv7 Processor rev 0 (v7l)
processor       : 1
BogoMIPS        : 2793.47
Features        : half fastmult edsp tls 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x3
CPU part        : 0xc09
CPU revision    : 0

Hardware        : Northstar Prototype
Revision        : 0000
Serial          : 0000000000000000

**Comment by xerial** [comment]
Related to #172. A problem is JVM does not distinguish various ARM types (v7l, etc.)

**Comment by sdelafond** [comment]
OK, but on the other hand the .so shipped in the 3.15.1 jar for armhf (./org/sqlite/native/Linux/armhf/libsqlitejdbc.so) isn't even being recognized as a valid dynamic executable by ldd, so I'm not sure what the resolution should be here. Are you saying that later on there should be yet another .so bundled in the jar, targetted for... what exactly ? :)

**Comment by xerial** [comment]
We need to have armv7l specific native library. uname -a can be used to distinguish architecture types at OSInfo.java

**Comment by sdelafond** [comment]
Well that's my point, I don't the problem lies there. I have another
arm-based box (Linksys wrt1900acs), that runs a Debian armhf userland
(enough flags in cpuinfo to allow for that), and jdbc-sqlite doesn't
crash on it.

Here are the specs, showing it's also a v7l.

# cat /proc/cpuinfo
model name      : ARMv7 Processor rev 1 (v7l)
processor       : 0
BogoMIPS        : 3170.30
Features        : half thumb fastmult vfp edsp vfpv3 tls vfpd32 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x4
CPU part        : 0xc09
CPU revision    : 1

model name      : ARMv7 Processor rev 1 (v7l)
processor       : 1
BogoMIPS        : 3194.88
Features        : half thumb fastmult vfp edsp vfpv3 tls vfpd32 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x4
CPU part        : 0xc09
CPU revision    : 1

Hardware        : Marvell Armada 380/385 (Device Tree)
Revision        : 0000
Serial          : 0000000000000000

# uname -a
Linux wrt1900acs.example.com 3.18.42 #1 SMP Thu Dec 1 11:40:35 CET 2016 armv7l GNU/Linux

**Comment by dmorris-untangle** [comment]
I worked on this also, and I'm pretty sure its not related to choosing the right .so file in OSInfo.java

To just eliminate this as an issue I pulled the source, did 'make native' on the target arch.
I then unjarred the distribution jar, and overwrote the arm and armhf .so file with the one for this arch and rejarred it.
For good measure I added a System.out.println to print what gets chosen in OSInfo.java
It was indeed loading the 'Linux/arm' .so file and the one that I compiled natively (the compile flags look correct FWIW).
Given all this it still crashes when opening the DB. Unfortunately, I don't see how/why it would crash in open_utf8.

Below is my test code. I've tried many different variations, but nothing I seem to do changes anything.
I tried an oracle VM and several openjdks.
The produce slightly different dump files, but they all show illegal instruction.
but its always in open_utf8

relevant oracle output:
j  org.sqlite.core.NativeDB._open_utf8([BI)V+0
j  org.sqlite.core.NativeDB._open(Ljava/lang/String;I)V+6
j  org.sqlite.core.DB.open(Lorg/sqlite/SQLiteConnection;Ljava/lang/String;I)V+8
j  org.sqlite.core.CoreConnection.open(II)V+408
...

relevant openjdk output:
Current thread (0x6b82b608):  JavaThread "UTThread-6" [_thread_in_native, id=30008, stack(0x65c40000,0x65c80000)]

Stack: [0x65c40000,0x65c80000],  sp=0x65c624ec,  free space=137k
Java frames:
 0x65c7e598: unboxed parameter[2]  = 0x65c7e608
 0x65c7e59c: unboxed parameter[1]  = 0x65c7e60c
 0x65c7e5a0: parameter[0] (JNIEnv) = 0x65c62da0
 0x65c7e5a4: parameter[1] (this)   = 0x65c7e59c
 0x65c7e5a8: parameter[2]          = 0x65c7e598
 0x65c7e5ac: parameter[3]          = 0x65c7e604
 0x65c7e5b0: monitor[0]->_lock     = 0x00000000
 0x65c7e5b4: monitor[0]->_obj      = 0x6c49ebf0
 0x65c7e5b8: istate->_thread       = 0x6b82b608
 0x65c7e5bc: istate->_bcp          = 0x00000000
 0x65c7e5c0: istate->_locals       = 0x65c7e60c
 0x65c7e5c4: istate->_constants    = 0x654f0018
 0x65c7e5c8: istate->_method       = org.sqlite.core.NativeDB._open_utf8([BI)V
 0x65c7e5cc: istate->_mdx          = 0x00000000
 0x65c7e5d0: istate->_stack        = 0x65c7e5ac
 0x65c7e5d4: istate->_msg          = 0x00000002
 0x65c7e5d8: istate->_result       = 0x00000000
 0x65c7e5dc: (istate->_result)     = 0x74400140
 0x65c7e5e0: (istate->_result)     = 0x00000003
 0x65c7e5e4: istate->_prev_link    = 0x00000000
 0x65c7e5e8: istate->_oop_temp     = 0x00000000
 0x65c7e5ec: istate->_stack_base   = 0x65c7e5b0
 0x65c7e5f0: istate->_stack_limit  = 0x65c7e5ac
 0x65c7e5f4: istate->_monitor_base = 0x65c7e5b8
 0x65c7e5f8: istate->_self_link    = 0x65c7e5b8
 0x65c7e5fc: frame_type            = INTERPRETER_FRAME
 0x65c7e600: next_frame            = 0x65c7e660
...

java test file:
package org.sqlite;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.Statement;
import java.sql.ResultSet;
import java.util.Properties;

public class Main
{
    public static void main(String[] args)
    {
        try {
        Properties prop = new Properties();
        //prop.setProperty("enable_load_extension", "true");

        Connection conn = null;
        Class.forName("org.sqlite.JDBC");
        System.out.println(Class.forName("org.sqlite.JDBC"));
        conn = DriverManager.getConnection("jdbc:sqlite:", prop);
        Statement stat = conn.createStatement();

            // How to build shared lib in Windows
            // # mingw32-gcc -fPIC -c extension-function.c
            // # mingw32-gcc -shared -Wl -o extension-function.dll extension-function.o

        ResultSet rs = stat.executeQuery("select count(*) from sessions");
        System.out.println(rs.getDouble(1));

        } catch (Exception e) {
        System.out.println(e);
        }

    }

    public static void majorVersion() throws Exception
    {
        int major = DriverManager.getDriver("jdbc:sqlite:").getMajorVersion();
        int minor = DriverManager.getDriver("jdbc:sqlite:").getMinorVersion();
        System.out.println("Major: " + major + "Minor: " + minor);
    }
}

**Comment by xerial** [comment]
For arm, sqlite uses softfloat compilation flag is added: 
```
Linux-arm_CCFLAGS     := -I$(JAVA_HOME)/include -Ilib/inc_linux -O2 -fPIC -mfloat-abi=softfp -mfpu=vfp -fvisibility=hidden
```

**Comment by xerial** [comment]
#193 could fix this problem.

**Comment by dmorris-untangle** [comment]
Thanks for the tip!

Its definitely related to these options -mfloat-abi=softfp -mfpu=vfp
I compiled the .so without those options and it does not crash. Progress!

I would try #193, but I can't get the full build working yet.

**Comment by xerial** [comment]
@dmorris-untangle please try this version:
https://oss.sonatype.org/content/repositories/snapshots/org/xerial/sqlite-jdbc/3.16.1-SNAPSHOT/sqlite-jdbc-3.16.1-20170106.232609-17.jar

**Comment by xerial** [comment]
Released 3.16.1 with a fix for this. Let me know if you still have any problem.

**Comment by sdelafond** [comment]
This is unfortunately still not fixed: we get the exact same crash as before. Looking at PR 192, it doesn't surprise me, as the cpuflags listed in /proc/cpuinfo are not taken into account anywhere: the absence/presence of vfp in those flags is IMHO the only way to properly fix this. Simply relying on armv7l being part of uname -a will not work.
